hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/AdjustmentUtils.java

```diff
@@ -0,0 +1,232 @@
+     * @param denominator The denominator of the fraction
+     * @param amount The given units transferred
+     * @return The amount to be charged
+     */
+    public static long safeFractionMultiply(final long numerator, final long denominator, final long amount) {
+        if (amount != 0 && numerator > Long.MAX_VALUE / amount) {
+            return BigInteger.valueOf(amount)
+                    .multiply(BigInteger.valueOf(numerator))
+                    .divide(BigInteger.valueOf(denominator))
+                    .longValueExact();
+        } else {
+            return numerator * amount / denominator;
+        }
+    }
+
+    /**
+     * Given the token deomination, fee collector and the amount to be charged, this method
+     * returns the {@link FixedFee} representation of custom fee.
+     * @param unitsToCollect The amount to be charged
+     * @param tokenDenomination The token denomination
+     * @param feeCollector The fee collector
+     * @param allCollectorsAreExempt Whether all collectors are exempt
+     * @return The {@link FixedFee} representation of custom fee
+     */
+    public static CustomFee asFixedFee(
+            final long unitsToCollect,
+            final TokenID tokenDenomination,
+            final AccountID feeCollector,
+            final boolean allCollectorsAreExempt) {
+        Objects.requireNonNull(feeCollector);
+        final var spec = FixedFee.newBuilder()
+                .denominatingTokenId(tokenDenomination)
+                .amount(unitsToCollect)
+                .build();
+        return CustomFee.newBuilder()
+                .fixedFee(spec)
+                .feeCollectorAccountId(feeCollector)
+                .allCollectorsAreExempt(allCollectorsAreExempt)
+                .build();
+    }
+
+    /**
+     * Adjusts a HTS fee. If the fee is self-denominated, it should not trigger custom fees again
+     * So add the adjustment to previous level transaction. If the fee is not self-denominated,
+     * it should trigger custom fees again. So add the adjustment to next level transaction.
+     * @param result The {@link AssessmentResult} object
+     * @param sender The sender account
+     * @param collector The fee collector
+     * @param chargingTokenMeta The {@link CustomFeeMeta} object of token to be charged
+     * @param amount The amount to be charged
+     * @param denominatingToken The token denomination
+     */
+    public static void adjustHtsFees(
+            final AssessmentResult result,
+            final AccountID sender,
+            final AccountID collector,
+            final CustomFeeMeta chargingTokenMeta,
+            final long amount,
+            final TokenID denominatingToken) {
+        final var newHtsAdjustments = result.getHtsAdjustments();
+        final var inputHtsAdjustments = result.getInputTokenAdjustments();
+
+        // If the fee is self-denominated, we don't need tit to trigger next level custom fees
```

Comment:
Typo, "We don't need ~~tit~~ it to trigger..."

Reply:
Oh joy !